File: kotlin-node/karakum/src/jsMain/resources/plugins/convertDotTestReporterPayload.js
import ts from "typescript";
import * as karakum from "../karakum.mjs"

export default function (node, context) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("test.d.ts")) return null

    if (
        (
            ts.isUnionTypeNode(node)

            && node.parent
            && ts.isTypeReferenceNode(node.parent)
            && ts.isIdentifier(node.parent.typeName)
            && node.parent.typeName.text === "AsyncGenerator"

            && node.parent.parent
            && ts.isFunctionDeclaration(node.parent.parent)
            && node.parent.parent.name?.text === "dot"
        )
    ) {
        const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())

        return `String /* ${typeScriptService?.printNode(node)} */`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertDotTestReporterPayload.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.isFunctionDeclaration
import typescript.isIdentifier
import typescript.isTypeReferenceNode
import typescript.isUnionTypeNode

val convertDotTestReporterPayload = createPlugin { node, context, _ ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("test.d.ts"))

        ensure(isUnionTypeNode(node))

        val typeReference = ensureNotNull(node.getParentOrNull())
        ensure(isTypeReferenceNode(typeReference))

        val typeReferenceName = typeReference.typeName
        ensure(isIdentifier(typeReferenceName))
        ensure(typeReferenceName.text == "AsyncGenerator")

        val function = ensureNotNull(typeReference.getParentOrNull())
        ensure(isFunctionDeclaration(function))

        val functionName = ensureNotNull(function.name)
        ensure(functionName.text == "dot")

        val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

        "String /* ${typeScriptService?.printNode(node)} */"
    }
}
